src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitWhileStatement(node);
+        }
+
+        public override BoundNode? VisitSwitchStatement(BoundSwitchStatement node)
+        {
+            using var _ = new LocalScope(this, node.InnerLocals);
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitSwitchStatement(node);
+            return null;
+        }
+
+        public override BoundNode? VisitConvertedSwitchExpression(BoundConvertedSwitchExpression node)
+        {
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitConvertedSwitchExpression(node);
+            return null;
+        }
+
+        public override BoundNode? VisitSwitchSection(BoundSwitchSection node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchSection(node);
+        }
+
+        public override BoundNode? VisitSwitchExpressionArm(BoundSwitchExpressionArm node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchExpressionArm(node);
+        }
+
+        public override BoundNode? VisitCatchBlock(BoundCatchBlock node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitCatchBlock(node);
+        }
+
+        public override BoundNode? VisitLocal(BoundLocal node)
+        {
+            Debug.Assert(_localEscapeScopes?.ContainsKey(node.LocalSymbol) == true ||
+                (node.LocalSymbol.ContainingSymbol is SynthesizedSimpleProgramEntryPointSymbol entryPoint && _symbol != entryPoint));
+
+            return base.VisitLocal(node);
+        }
+
+        private void AddLocalScopes(LocalSymbol local, uint refEscapeScope, uint valEscapeScope)
+        {
+            var scope = _useUpdatedEscapeRules ? local.Scope : DeclarationScope.Unscoped;
+            if (scope != DeclarationScope.Unscoped)
+            {
+                // From https://github.com/dotnet/csharplang/blob/main/csharp-11.0/proposals/low-level-struct-improvements.md:
+                //
+                // | Parameter or Local     | ref-safe-to-escape | safe-to-escape |
+                // |------------------------|--------------------|----------------|
+                // | Span<int> s            | current method     | calling method |
+                // | scoped Span<int> s     | current method     | current method |
+                // | ref Span<int> s        | calling method     | calling method |
+                // | scoped ref Span<int> s | current method     | calling method |
+
+                refEscapeScope = scope == DeclarationScope.RefScoped ?
+                    _localScopeDepth :
+                    Binder.CurrentMethodScope;
```

Comment:
Not related to current PR: I couldn't follow this logic in relation to the table above.

Reply:
> I think the code would make more sense if we did the following. ...

The current implementation should match the previous behavior from `SourceLocalSymbol.RefEscapeScope` and `SourceLocalSymbol.ValEscapeScope`. 

I prefer your suggestion but that appears to change the behavior of at least one test, `RefFieldTests.LocalScope_11_Foreach_02`, so I'd like to investigate separately.

Reply:
[https://github.com/dotnet/roslyn/issues/66079](https://github.com/dotnet/roslyn/issues/66079)